File: src/C/lu_matrix.c
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-21 14:13:48
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-23 16:59:31
@FilePath: \asm_matrix_benchmark\src\C\lu_matrix.c
@Description: lu decomposition of matrix c code
*************************************************************/
#include "lu_matrix.h"

/***********************************************************
@description: lu decomposition of matrix
@param {MatrixInt} *m
@param {LU_Result} *res
@return {*}
 ************************************************************/
bool LU_Decomposition_int(MatrixInt *m, LU_Result *res)
{
    // check m and m->data
    if (m == NULL || m->data == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return false;
    }
    // check dimension
    if (m->cols != m->rows)
    {
        puts("It's not a square");
        return false;
    }
    // check the matrix can lu decomposition or not
    Leading_Minors_Int *leading_minors = get_leading_minors_int(m);
    if (leading_minors == NULL)
    {
        fprintf(stderr, "function get_leading_minors_int has a wrong value\n");
        return NULL;
    }
    for (size_t i = 0; i < leading_minors->len; i++)
    {
        int rank = 0;
        if (rank_matrix_int(leading_minors->matrix_data + i, &rank) == false)
        {
            return false;
        }
        else
        {
            if ((size_t)rank != i + 1)
            {
                fprintf(stderr, "This matrix can't lu decomposition\n");
                return false;
            }
        }
    }

    // free leading_minors
    for (size_t i = 0; i < leading_minors->len; i++)
    {
        free((leading_minors->matrix_data + i)->data);
    }
    free(leading_minors->matrix_data);
    free(leading_minors);

    // malloc new L
    MatrixDouble *L = NULL;
    L = (MatrixDouble *)malloc(sizeof(MatrixDouble));
    if (L == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return false;
    }
    L->rows = m->rows;
    L->cols = m->cols;
    // malloc new L data
    L->data = (double *)malloc(sizeof(double) * L->rows * L->cols);
    if (L->data == NULL)
    {
        free(L);
        fprintf(stderr, "Memory allocation failed\n");
        return false;
    }
    // malloc new U
    MatrixDouble *U = NULL;
    U = (MatrixDouble *)malloc(sizeof(MatrixDouble));
    if (U == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return false;
    }
    U->rows = m->rows;
    U->cols = m->cols;
    // malloc new U data
    U->data = (double *)malloc(sizeof(double) * U->rows * U->cols);
    if (U->data == NULL)
    {
        free(U);
        fprintf(stderr, "Memory allocation failed\n");
        return false;
    }
    // malloc new res
    if (res == NULL)
    {
        res = (LU_Result *)malloc(sizeof(LU_Result));
        if (res == NULL)
        {
            fprintf(stderr, "Memory allocation failed\n");
            return false;
        }
    }

    // The specific process of LU decomposition is as follows:
    /*
    |a_{11} & a_{12} & a_{13} & \cdots & a_{1n}|
    |a_{21} & a_{22} & a_{23} & \cdots & a_{2n}|
    |a_{31} & a_{32} & a_{33} & \cdots & a_{3n}|
    |\vdots & \vdots & \vdots & \ddots & \vdots|
    |a_{n1} & a_{n2} & a_{n3} & \cdots & a_{nn}|

    =

    |1      &        &        &        &       |
    |l_{21} & 1      &        &        &       |
    |l_{31} & l_{32} & 1      &        &       |
    |\vdots & \vdots & \vdots & \ddots &       |
    |l_{n1} & l_{n2} & l_{n3} & \cdots & 1     |

    \*

    |u_{11} & u_{12} & u_{13} & \cdots & u_{1n}|
    |       & u_{22} & u_{23} & \cdots & u_{2n}|
    |       &        & u_{33} & \cdots & u_{3n}|
    |       &        &        & \ddots & \vdots|
    |       &        &        &        & u_{nn}|
    */

    // From matrix multiplication, we obtain:
    // init L matrix upper triangle
    for (size_t i = 0; i < L->rows; i++)
    {
        for (size_t j = i; j < L->cols; j++)
        {
            L->data[i * L->cols + j] = (i == j ? 1.0 : 0.0);
        }
    }

    // init U matrix lower triangle
    for (size_t i = 0; i < U->rows; i++)
    {
        for (size_t j = 0; j < i; j++)
        {
            U->data[i * U->cols + j] = 0.0;
        }
    }

    /*
    u_{1j} = a_{1j},  j = 1:n
    l_{i1} = a_{i1} / u_{11},  i = 2:n
    */
    for (size_t i = 0; i < U->cols; i++)
    {
        U->data[i] = m->data[i];
    }
    for (size_t i = 1; i < L->rows; i++)
    {
        L->data[i * L->cols + 0] = m->data[i * m->cols + 0] / U->data[0 * U->cols + 0];
    }

    /*
    u_{ij} = a_{ij} - sum_{k=1}^{i-1} l_{ik} u_{kj},  j = i:n
    l_{ij} = (a_{ij} - sum_{k=1}^{j-1} l_{ik} u_{kj}) / u_{jj},  i = j+1:n
    */
    for (size_t i = 1; i < L->rows; i++)
    {
        for (size_t j = i; j < L->rows; j++)
        {
            double sum = 0;
            for (size_t k = 0; k < i; k++)
            {
                sum += (L->data[i * L->cols + k] * U->data[k * U->cols + j]);
            }
            U->data[i * U->cols + j] = m->data[i * m->cols + j] - sum;
        }
        for (size_t j = i + 1; j < L->rows; j++)
        {
            double sum = 0;
            for (size_t k = 0; k < j; k++)
            {
                sum += (L->data[j * L->cols + k] * U->data[k * U->cols + i]);
            }
            L->data[j * L->cols + i] = (m->data[j * m->cols + i] - sum) / U->data[i * U->cols + i];
        }
    }
    // final result
    res->L = L;
    res->U = U;
    return true;
}

File: src/JavaScript/lu_matrix.js
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-29 20:10:32
@LastEditors: GitHubCopilot
@LastEditTime: 2026-06-29 20:10:32
@FilePath: \asm_matrix_benchmark\src\JavaScript\lu_matrix.js
@Description: LU decomposition JavaScript code
*************************************************************/

"use strict";

const {
  checkMatrix,
  buildTypedMatrix,
  EPS_FLOAT,
  EPS_DOUBLE,
} = require("./common");
const {
  determinant_int,
  determinant_float,
  determinant_double,
} = require("./determinant_matrix");
const {
  get_leading_minors_int,
  get_leading_minors_float,
  get_leading_minors_double,
} = require("./leading_minors");

/**
 * Check if all leading principal minors are non-zero.
 * This follows the C implementation logic before Doolittle decomposition.
 * @param {{rows:number, cols:number, data:number[]}} m
 * @param {"int"|"float"|"double"} type
 * @returns {boolean}
 */
function hasNonZeroLeadingMinors(m, type) {
  const minors =
    type === "int"
      ? get_leading_minors_int(m)
      : type === "float"
        ? get_leading_minors_float(m)
        : get_leading_minors_double(m);

  if (minors === null) {
    return false;
  }

  for (let i = 0; i < minors.length; i++) {
    const det =
      type === "int"
        ? determinant_int(minors[i])
        : type === "float"
          ? determinant_float(minors[i])
          : determinant_double(minors[i]);

    if (det === null) {
      return false;
    }

    const eps = type === "int" ? 0 : type === "float" ? EPS_FLOAT : EPS_DOUBLE;
    if (Math.abs(det) <= eps) {
      return false;
    }
  }

  return true;
}

/**
 * Doolittle LU decomposition without pivoting.
 * Output matrices use double flavor for numerical stability,
 * consistent with C code where LU outputs MatrixDouble.
 * @param {{rows:number, cols:number, data:number[]}} m
 * @param {"int"|"float"|"double"} type
 * @returns {{L:{rows:number, cols:number, data:number[]}, U:{rows:number, cols:number, data:number[]}}|null}
 */
function luDecomposition(m, type) {
  if (!checkMatrix(m, "m")) {
    return null;
  }
  if (m.rows !== m.cols) {
    console.error("Matrix must be square!");
    return null;
  }

  if (!hasNonZeroLeadingMinors(m, type)) {
    console.error("Matrix cannot perform LU decomposition without pivoting!");
    return null;
  }

  const n = m.rows;
  const L = new Array(n * n).fill(0);
  const U = new Array(n * n).fill(0);

  for (let i = 0; i < n; i++) {
    L[i * n + i] = 1;
  }

  // Doolittle formula: solve U row i first, then L column i.
  for (let i = 0; i < n; i++) {
    for (let k = i; k < n; k++) {
      let sum = 0;
      for (let j = 0; j < i; j++) {
        sum += L[i * n + j] * U[j * n + k];
      }
      U[i * n + k] = m.data[i * n + k] - sum;
    }

    for (let k = i + 1; k < n; k++) {
      let sum = 0;
      for (let j = 0; j < i; j++) {
        sum += L[k * n + j] * U[j * n + i];
      }
      U[i * n + i] = Number(U[i * n + i]);
      if (Math.abs(U[i * n + i]) <= EPS_DOUBLE) {
        console.error("Singular pivot found in LU decomposition!");
        return null;
      }
      L[k * n + i] = (m.data[k * n + i] - sum) / U[i * n + i];
    }
  }

  return {
    L: buildTypedMatrix(n, n, L, "double"),
    U: buildTypedMatrix(n, n, U, "double"),
  };
}

function LU_Decomposition_int(m) {
  return luDecomposition(m, "int");
}

function LU_Decomposition_float(m) {
  return luDecomposition(m, "float");
}

function LU_Decomposition_double(m) {
  return luDecomposition(m, "double");
}

module.exports = {
  LU_Decomposition_int,
  LU_Decomposition_float,
  LU_Decomposition_double,
};
